jujugui/static/gui/src/app/state/test-state.js

```diff
@@ -0,0 +1,566 @@
+      const state = new window.jujugui.State({
+        baseURL: 'http://abc.com:123'
+      });
+
+      const urls = [{
+        path: 'http://abc.com:123/new',
+        state: { root: 'new' }
+      },{
+        path: 'http://abc.com:123/store',
+        state: { root: 'store' }
+      },{
+        path: 'http://abc.com:123/about',
+        state: { root: 'about' }
+      },{
+        path: 'http://abc.com:123/docs',
+        state: { root: 'docs' }
+      },{
+        path: 'http://abc.com:123/login',
+        state: { root: 'login' }
+      }];
+
+      urls.forEach(test => {
+        assert.deepEqual(
+          state.buildState(test.path),
+          test.state,
+          `${test.path} did not properly generate the state object: ` +
+          JSON.stringify(test.state));
+      });
+    });
+
+    it('builds the proper state for the search urls', () => {
+      const state = new window.jujugui.State({
+        baseURL: 'http://abc.com:123'
+      });
+
+      const urls = [{
+        path: 'http://abc.com:123/q/haproxy',
+        state: { search: 'haproxy' }
+      }, {
+        path: 'http://abc.com:123/q/k8s/core',
+        state: { search: 'k8s/core' }
+      }];
+
+      urls.forEach(test => {
+        assert.deepEqual(
+          state.buildState(test.path),
+          test.state,
+          `${test.path} did not properly generate the state object: ` +
+          JSON.stringify(test.state));
+      });
+    });
+
+    it('builds the proper state for the gui urls', () => {
+      const state = new window.jujugui.State({
+        baseURL: 'http://abc.com:123'
+      });
+
+      const urls = [{
+        path: 'http://abc.com:123/i/inspector/haproxy/config',
+        state: { gui: { inspector: 'haproxy/config' }}
+      }, {
+        path: 'http://abc.com:123/i/machines',
+        state: { gui: { machines: '' }}
+      }, {
+        path: 'http://abc.com:123/i/applications/inspector/ghost',
+        state: { gui: { applications: '', inspector: 'ghost' }}
+      }, {
+        path:
+          'http://abc.com:123/i/inspector/apache2/machines/3/lxc-0/deploy/foo',
+        state: {
+          gui: { inspector: 'apache2', machines: '3/lxc-0', deploy: 'foo'}
+        }
+      }];
+
+      urls.forEach(test => {
+        assert.deepEqual(
+          state.buildState(test.path),
+          test.state,
+          `${test.path} did not properly generate the state object: ` +
+          JSON.stringify(test.state));
+      });
+    });
+
+    it('builds proper state for the user urls', () => {
+      const state = new window.jujugui.State({
+        baseURL: 'http://abc.com:123'
+      });
+
+      const urls = [{
+        path: 'http://abc.com:123/u/ant',
+        state: { user: 'ant' }
+      }, {
+        path: 'http://abc.com:123/u/hatch/staging',
+        state: { user: 'hatch/staging' }
+      }, {
+        path: 'http://abc.com:123/u/hatch/mongodb/xenial',
+        state: { store: 'hatch/mongodb/xenial' }
```

Comment:
Wouldn't it be great if the store field holds a jujulib.URL instance rather than just a string?